src/ripple/legacy/0.27/CreateOffer.cpp

```diff
@@ -0,0 +1,631 @@
+        // can reduce the size of the order book as much as possible
+        // before terminating the loop.
+
+        if (taker.done())
+        {
+            m_journal.debug << "The taker reports he's done during crossing!";
+            break;
+        }
+
+        if (! offers.step ())
+        {
+            // Place the order since there are no
+            // more offers and the order has a balance.
+            m_journal.debug << "No more offers to consider during crossing!";
+            break;
+        }
+
+        auto const& offer (offers.tip());
+
+        if (taker.reject (offer.quality()))
+        {
+            // Place the order since there are no more offers
+            // at the desired quality, and the order has a balance.
+            break;
+        }
+
+        if (offer.account() == taker.account())
+        {
+            // Skip offer from self. The offer will be considered expired and
+            // will get deleted.
+            continue;
+        }
+
+        if (m_journal.debug) m_journal.debug <<
+                "  Offer: " << offer.entry()->getIndex() << std::endl <<
+                "         " << offer.amount().in << " : " << offer.amount().out;
+
+        cross_result = taker.cross (offer);
+
+        if (cross_result != tesSUCCESS)
+        {
+            cross_result = tecFAILED_PROCESSING;
+            break;
+        }
+    }
+
+    return std::make_pair(cross_result, taker.remaining_offer ());
+}
+
+CreateOffer::CreateOffer (STTx const& txn,
+    TransactionEngineParams params, TransactionEngine* engine)
+    : Transactor (txn, params, engine, deprecatedLogs().journal("CreateOffer"))
+{
+}
+
+TER
+CreateOffer::doApply()
+{
+    if (m_journal.debug) m_journal.debug <<
+        "OfferCreate> " << mTxn.getJson (0);
+
+    std::uint32_t const uTxFlags = mTxn.getFlags ();
+
+    bool const bPassive (uTxFlags & tfPassive);
+    bool const bImmediateOrCancel (uTxFlags & tfImmediateOrCancel);
+    bool const bFillOrKill (uTxFlags & tfFillOrKill);
+    bool const bSell  (uTxFlags & tfSell);
+
+    STAmount saTakerPays = mTxn.getFieldAmount (sfTakerPays);
+    STAmount saTakerGets = mTxn.getFieldAmount (sfTakerGets);
+
+    if (!isLegalNet (saTakerPays) || !isLegalNet (saTakerGets))
+        return temBAD_AMOUNT;
+
+    auto const& uPaysIssuerID = saTakerPays.getIssuer ();
+    auto const& uPaysCurrency = saTakerPays.getCurrency ();
+
+    auto const& uGetsIssuerID = saTakerGets.getIssuer ();
+    auto const& uGetsCurrency = saTakerGets.getCurrency ();
+
+    bool const bHaveExpiration (mTxn.isFieldPresent (sfExpiration));
+    bool const bHaveCancel (mTxn.isFieldPresent (sfOfferSequence));
+
+    std::uint32_t const uExpiration = mTxn.getFieldU32 (sfExpiration);
+    std::uint32_t const uCancelSequence = mTxn.getFieldU32 (sfOfferSequence);
+
+    // FIXME understand why we use SequenceNext instead of current transaction
+    //       sequence to determine the transaction. Why is the offer seuqnce
```

Comment:
s/seuqnce/sequence/

But actually - I think we should clear this up before you check in if we can.